### DISCUSSION TITLE
Can I pre-set or calculate the cbRequested that I can receive from 'ServerFileContentsRequest'?

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#8860] [Q&A] [answered]
Hello, thanks to your kind help, the file clipboard implementation is complete.

`ServerFileContentsRequest` sends data as much as `fileContentsRequest->cbRequested` when `fileContentsRequest->dwFlags` is `FILECONTENTS_RANGE`, where `fileContentsRequest->cbRequested` depends on the size of the file.

Q. I can set up `cbRequested` when I need to get `fileContents` from the client, but is there a way to set up `cbRequested` when I need to send it?
Is there a formula where `cbRequested` is set separately if it can't be set?

I'm asking because the implementation method of my project needs to be changed depending on whether the cbRequest can be set to a specific value.

**Comment by akallabeth** [comment]
@kimcoldwater sorry, but what?
This sounds like you still do not understand how the whole process works.
1. If you do not know the remote file size send a request with `FILECONTENTS_SIZE` flag set, the response will be the file size. (check existing file clipboard implementations for details)
2. If you know the size of the file you request then you already know it and your question is rendered moot
3. If you need something more specific and can not do that in `java/lotlin` write your own `JNI` wrapper just like we did with `aFreeRDP` (you are still on that project, right?)

**Reply by kimcoldwater** [reply]
@akallabeth  

I don't mean when I receive a remote file.

When sending local files to a remote session.

With reference to `xf_cliprdr_server_file_content_request` and `wlf_cliprdr_server_file_contents_request`, 
I implemented processing when dwFlags is `FILECONTENTS_SIZE` and `FILECONTENTS_RANGE`.

When `FILECONTENTS_SIZE`, I sent the file size in `fileContentsResponse.requestedData`,
`FILECONTENTS_RANGE` implemented the file clipboard by sending file data as many as `fileContentsRequest->cbRequested` and `fileContentsRequest->nPosition`.

The server asked if the requested `fileContentsRequested->cbRequested` could be set to a specific value and if there is a specific formula to be calculated.

**Reply by pnowack** [reply, marked as answer]
> The server asked if the requested fileContentsRequested->cbRequested could be set to a specific value and if there is a specific formula to be calculated.

Is your question, how the other peer determines how many bytes it requests? If yes, then that would depend on the server side.
In the case of MS Windows RDS, you normally always get requests of 256KiB or 256KB (don't remember whether that was an IEC or SI unit, what MS Windows requests in size).
This is regardless of the actual file size. If e.g. a file has the filesize of 45KiB, then you likely still get a request of 256KiB. In that case, you just deliver the 45KiB, because that's it.
If the filesize is e.g. 30MiB, then you can expect, multiple requests (e.g. each of them requesting 256KiB at an incrementing offset). The last request MAY go over the 30MiB like in the 45KiB, but that's okay.

256KiB or 256KB (don't remember whether that was an IEC or SI unit, what MS Windows requests in size) is the typical size with MS Windows RDS.
On other servers, if FUSE is used, then without direct_io (in the FUSE config on the other peer), the request size is typically 128KiB, with direct_io 1024KiB, afair.

**Reply by pnowack** [reply]
But in any case, you have to expect that the server side (or client side, depending whether you implement you clipboard interface for a client or server) MAY choose some arbitrary value of the amount of bytes it requests for a file at an offset.

It could just request 200KiB of a file, then maybe another slice of 300KiB, then maybe just 50KiB.

**Reply by kimcoldwater** [reply]
Thank you for your kind reply.
Your help is always a great help to me. 
:)
